staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache_interval_test.go

```diff
@@ -0,0 +1,417 @@
+	}
+
+	for _, c := range cases {
+		wcib := &watchCacheIntervalBuffer{endIndex: c.endIndex}
+		actual := wcib.isFull()
+		if actual != c.expected {
+			t.Errorf("expected %v, got %v", c.expected, actual)
+		}
+	}
+}
+
+func TestIntervalBufferIsEmpty(t *testing.T) {
+	cases := []struct {
+		startIndex int
+		endIndex   int
+		expected   bool
+	}{
+		{startIndex: 0, endIndex: 10, expected: false},
+		{startIndex: 5, endIndex: 20, expected: false},
+		{startIndex: 50, endIndex: 50, expected: true},
+	}
+
+	for _, c := range cases {
+		wcib := &watchCacheIntervalBuffer{
+			startIndex: c.startIndex,
+			endIndex:   c.endIndex,
+		}
+		actual := wcib.isEmpty()
+		if actual != c.expected {
+			t.Errorf("expected %v, got %v", c.expected, actual)
+		}
+	}
+}
+
+func TestIntervalBufferNext(t *testing.T) {
+	cases := []struct {
+		events []*watchCacheEvent
+	}{
+		{
+			events: []*watchCacheEvent{
+				{Type: watch.Added, Object: makeTestPod("pod1", 1)},
+				{Type: watch.Added, Object: makeTestPod("pod2", 2)},
+				{Type: watch.Modified, Object: makeTestPod("pod3", 3)},
+			},
+		},
+		{
+			events: []*watchCacheEvent{},
+		},
+	}
+
+	for _, c := range cases {
+		wcib := bufferFromEvents(c.events)
+
+		received := 0
+		for {
+			event, ok := wcib.next()
+			if !ok {
+				if received != len(c.events) {
+					t.Errorf("unexpected bool value indicating empty buffer")
+				}
+				break
+			}
+			if received == len(c.events) {
+				t.Errorf("unexpected bool value indicating non-empty buffer")
+				break
+			}
+			if event == nil {
+				t.Errorf("unexpected nil event, expected: %v", *c.events[received])
+				break
+			}
+			if !reflect.DeepEqual(event, c.events[received]) {
+				t.Errorf("unexpected event, expected: %v, got %v", *c.events[received], *event)
+			}
+			received++
+		}
+		event, ok := wcib.next()
+		if ok {
+			t.Error("unexpected bool value indicating buffer is not empty")
+		}
+		if event != nil {
+			t.Errorf("unexpected event received, expected: nil, got %v", *event)
+		}
+	}
+}
+
+func TestFillBuffer(t *testing.T) {
+	min := func(a, b int) int {
+		if a < b {
+			return a
+		}
+		return b
+	}
+
+	cases := []struct {
+		numEventsToFill int
+	}{
+		{numEventsToFill: 0},              // no events to put in buffer
+		{numEventsToFill: 5},              // some events to put in buffer < bufferSize
+		{numEventsToFill: bufferSize},     // some events to put in buffer = bufferSize
+		{numEventsToFill: bufferSize + 5}, // some events to put in buffer > bufferSize
+	}
+	for _, c := range cases {
+		events := generateEvents(0, c.numEventsToFill)
+		wci := intervalFromEvents(events)
+
+		eventsProcessed, eventsToProcess := 0, len(events)
+		for eventsToProcess > 0 {
+			wci.fillBuffer()
+			// In one cycle of fillBuffer() you can only receive
+			// bufferSize number of events at the maximum.
+			eventsToRecv := min(bufferSize, eventsToProcess)
+			received := 0
+			for {
+				event, ok := wci.buffer.next()
+				if !ok {
```

Comment:
it seems the code here is almost a copy-paste from the test above.
I recommend extracting it to a separate helper function, sth like:
```
func verifyEvent(t *testing.T, event, ok, expectedEvent) {
  // roughlty the copy paste of what you have here
}

func verifyNoEvent(t *testing.T, event, ok) {
  if ok {
    t.Errorf("...")
 }
}
```

and using these helper function in all tests in this file
``

Reply:
I had done this initially, but went with the current approach because logging would not provide accurate line numbers. But considering `t.Run(..)` is going to be used and the test cases named, I'll move back to this approach!

Reply:
Yeah - t.Run does the trick you need.